clarity/src/vm/analysis/type_checker/mod.rs

```diff
@@ -452,6 +466,39 @@ impl FunctionType {
     }
 }
 
+fn check_function_arg_signature<T: CostTracker>(
+    cost_tracker: &mut T, 
+    expected_sig: &FunctionArgSignature,
+     actual_type: &TypeSignature) -> CheckResult<()> 
+{
+    match expected_sig {
+        FunctionArgSignature::Single(expected_type) => {
+            analysis_typecheck_cost(cost_tracker, expected_type, actual_type)?;
+            if !expected_type.admits_type(actual_type) {
+                return Err(CheckErrors::TypeError(
+                    expected_type.clone(), 
+                    actual_type.clone()
+                ).into());
+            }
+        },
+        FunctionArgSignature::Union(expected_types) => {
+            let mut admitted = false;
+            for expected_type in expected_types.iter() {
+                analysis_typecheck_cost(cost_tracker, expected_type, actual_type)?;
+                if expected_type.admits_type(actual_type) {
+                    admitted = true;
+                    break;
```

Comment:
You might as well just error out right here, instead of having an extra `admitted` flag.

Reply:
It's not necessarily the first expected_type that matches the actual type.  Hence my earlier comment on "should a caller really have to pay extra using a method that accepts multiple possible types?" (which is a decision made by "us", not the caller).  Erroring out here would return if the first "union type" didn't match.